src/cmd/linuxkit/push_openstack.go

```diff
@@ -0,0 +1,251 @@
+		flags.Usage()
+		os.Exit(1)
+	}
+	filePath := remArgs[0]
+	// Check that the file both exists, and can be read
+	checkFile(filePath)
+
+	var data KeyStoneV3
+	// Defaulting to password login, other login methods may be added later
+	data.Auth.Identity.Methods = append(data.Auth.Identity.Methods, "password")
+
+	data.Auth.Identity.Password.User.Name = *usernameFlag
+	data.Auth.Identity.Password.User.Domain.Name = *userDomainFlag
+	data.Auth.Identity.Password.User.Password = *passwordFlag
+	data.Auth.Scope.Project.Domain.Name = *projectDomain
+	data.Auth.Scope.Project.Name = *projectName
+
+	payloadBytes, err := json.Marshal(data)
+	if err != nil {
+		log.Fatalf("Error building JSON\n%v", err)
+	}
+	body := bytes.NewReader(payloadBytes)
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/v3/auth/tokens", *keystoneAddress), body)
+	if err != nil {
+		log.Fatalf("Error Creating HTTP Request\n%v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		log.Fatalf("Return Code: %s\nError:%v", resp.Status, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == 201 { // OK
+		// Find the OpenStack KeyStone Token
+		openstackToken := resp.Header.Get("x-subject-token")
+
+		log.Debugf("Token => %s", openstackToken)
+
+		if openstackToken == "" {
+			log.Fatalln("Error: Can't locate OpenStack Token in Headers")
+		}
+		// Create a new Image (which will be left in a queued state)
+		imageID := createOpenStackImage(filePath, *imageName, *glanceAddress, openstackToken)
+		// Take the returned ImageID and upload our image to the created ID
+		uploadOpenStackImage(filePath, *glanceAddress, openstackToken, imageID)
+	} else {
+		message, _ := ioutil.ReadAll(resp.Body)
+		log.Fatalf("Error authenticating with OpenStack, Error Details:\n%s", string(message))
+	}
+
+}
+
+func createOpenStackImage(filePath string, name string, glanceAddress string, token string) string {
+	// Currently supported image formats that are both supported by LinuxKit and OpenStack Glance V2
+	formats := []string{"ami", "vhd", "vhdx", "vmdk", "raw", "qcow2", "iso"}
+
+	// Find extension of the filename and remove the leading stop
+	fileExtension := strings.Replace(path.Ext(filePath), ".", "", -1)
+	fileName := strings.TrimSuffix(path.Base(filePath), filepath.Ext(filePath))
+	// Check for Supported extension
+	var supportedExtension bool
+	supportedExtension = false
+	for i := 0; i < len(formats); i++ {
+		if strings.ContainsAny(fileExtension, formats[i]) {
+			supportedExtension = true
+		}
+	}
+
+	if supportedExtension == false {
+		log.Fatalf("Extension [%s] is not supported", fileExtension)
+	}
+
+	var image GlanceV2Image
+	image.ContainerFormat = "bare"
+	image.DiskFormat = fileExtension
+	if name == "" {
+		image.Name = fileName
+	} else {
+		image.Name = name
+	}
+
+	payloadBytes, err := json.Marshal(image)
+	if err != nil {
+		log.Fatalf("Error building JSON\n%v", err)
```

Comment:
generally we dont put the error reported on a new line, just use `blah: %v` elsewhere